### PR TITLE
Deletepostdiffs

### PR DESCRIPTION
**Pull request by gasoved**
#9109

### CODE REVIEW

File: src/controllers/write/posts.js [hunk 1]
```diff
@@ -94,3 +95,19 @@ Posts.restoreDiff = async (req, res) => {
 	helpers.formatApiResponse(200, res, await api.posts.restoreDiff(req, { ...req.params }));
 };
 
+Posts.deleteDiff = async (req, res) => {
+	if (!parseInt(req.params.pid, 10)) {
+		throw new Error('[[error:invalid-data]]');
+	}
+
+	const cid = await posts.getCidByPid(req.params.pid);
+	const isModerator = privileges.users.isModerator(cid, req.uid);
```
Comment (julianlam) [inline src/controllers/write/posts.js:104]: Perhaps use `user.isPrivileged` here instead, so you can also capture admins and gmods.

File: public/src/client/topic/diffs.js [hunk 2]
```diff
@@ -23,64 +61,50 @@ define('forum/topic/diffs', ['api', 'forum/topic/images'], function (api) {
...
-					title: '[[topic:diffs.title]]',
-					message: html,
-					size: 'large',
-				});
+				const modalExists = !!modal;
+				if (modalExists) {
+					modal.find('.modal-body').html(html);
+				} else {
+					modal = bootbox.dialog({ title: '[[topic:diffs.title]]', message: html, size: 'large' });
+				}
```
Comment (julianlam) [inline public/src/client/topic/diffs.js:71]: I am not sold on this, I will propose an alternative.

File: src/controllers/write/posts.js [hunk 3]
```diff
@@ -94,3 +95,19 @@ Posts.restoreDiff = async (req, res) => {
...
+
+	if (!isModerator) {
+		return helpers.formatApiResponse(403, res, new Error('[[error:no-privileges]]'));
+	}
+
+	await posts.diffs.delete(req.params.pid, req.params.timestamp, req.uid);
+
+	helpers.formatApiResponse(200, res);
```
Comment (julianlam) [inline src/controllers/write/posts.js:112]: Regarding the comment above re: `openModal`, have this API call `Diffs.get()` again, and return the result.

**Review by julianlam — CHANGES_REQUESTED** [review]

**Comment by julianlam** [comment]
There is a helper method client-side called `app.parseAndTranslate`, which takes arguments `(template, blockName, data, callback)`

The `blockName` argument is the one that is interesting, because if provided, the method will only parse and return the specific block in the template.

So, instead of re-opening the modal and parsing everything again, just call `app.parseAndTranslate('partials/modals/post_history', 'diffs', diffs, next);`

Where `diffs` is an array of diffs, the result of `Diffs.get()` server-side.

On your deletion API call, have it return the result of Diffs.get(), and basically pipe the result into this client-side method to regenerate the html.

Then replace the `<select>` with the new elements :smile:

File: src/controllers/write/posts.js (hunk 1 above)
Comment (gasoved) [reply src/controllers/write/posts.js]: Fixed, my corrupted installation mislead me. #isModerator captures gmods and #isPrivileged includes mods of **any** category, so I end up adding just #isAdministrator. This also led to fixing some existing bugs.

File: public/src/client/topic/diffs.js (hunk 2 above)
Comment (gasoved) [reply public/src/client/topic/diffs.js]: Number of diffs was also required to be updated, so I went with that ugly solution. Pushed a better one. :)

File: src/controllers/write/posts.js (hunk 1 above)
Comment (julianlam) [reply src/controllers/write/posts.js]: Okay, that's fine. We may want to revisit this and allow gmods access as well, but perhaps we can just make it a privilege.

Later :smile:

**Comment by pitaj** [comment]
> So, instead of re-opening the modal and parsing everything again, just call `app.parseAndTranslate('partials/modals/post_history', 'diffs', diffs, next);`

Or you can call benchpress directly

**Comment by julianlam** [comment]
Yes, but where's the fun in that :wink: Also it doesn't translate, no?

**Comment by julianlam** [comment]
@gasoved Let's continue this in #9241

File: src/controllers/write/posts.js (hunk 1 above)
Comment (gasoved) [reply src/controllers/write/posts.js]: Sure (It allows gmods access though).

**Pull request closed without merging**